src/librustc/mir/tcx.rs

```diff
@@ -110,49 +111,102 @@ impl<'tcx> Place<'tcx> {
     /// be `self` in the current MIR, because that is the only time we directly access the fields
     /// of a closure type.
-    pub fn is_upvar_field_projection<'cx, 'gcx>(&self, mir: &'cx Mir<'tcx>,
-                                                tcx: &TyCtxt<'cx, 'gcx, 'tcx>) -> Option<Field> {
-        let place = if let Place::Projection(ref proj) = self {
-            if let ProjectionElem::Deref = proj.elem {
-                &proj.base
+    pub fn is_upvar_field_projection<'cx, 'gcx>(
+        &self,
+        mir: &'cx Mir<'tcx>,
+        tcx: &TyCtxt<'cx, 'gcx, 'tcx>,
+    ) -> Option<Field> {
+        let base_place;
+        let mut place = self;
+        let mut by_ref = false;
+
+        base_place = place.base_place(*tcx);
+
+        if let Some(ProjectionElem::Deref) = place.projection() {
+            place = &base_place;
+            by_ref = true;
+        }
+        if let Some(ProjectionElem::Field(field, _ty)) = place.projection() {
+            let base_ty = place.base_place(*tcx).ty(mir, *tcx).to_ty(*tcx);
+
+            if base_ty.is_closure() || base_ty.is_generator()
+                && (!(by_ref && !mir.upvar_decls[field.index()].by_ref)) {
+                Some(*field)
             } else {
-                self
+                None
             }
         } else {
-            self
-        };
-
-        match place {
-            Place::Projection(ref proj) => match proj.elem {
-                ProjectionElem::Field(field, _ty) => {
-                    let base_ty = proj.base.ty(mir, *tcx).to_ty(*tcx);
+            None
+        }
+    }
 
-                    if  base_ty.is_closure() || base_ty.is_generator() {
-                        Some(field)
-                    } else {
-                        None
-                    }
+    // for Place:
+    //     Base.[a, b, c]
+    //                 ^-- projection
+    // => (Base.[a, b], c])
+    //     ^^^^^^^^^^   ^-- projection
+    //     |-- base_place
+    pub fn split_projection<'cx, 'gcx>(
+        &self,
+        tcx: TyCtxt<'cx, 'gcx, 'tcx>,
+    ) -> Option<(Place<'tcx>, &PlaceElem<'tcx>)> {
+        // split place_elems
+        // Base.[a, b, c]
+        //       ^^^^  ^-- projection(projection lives in the last elem)
+        //       |-- place_elems
+        match self.elems.split_last() {
+            Some((projection, place_elems)) => Some((
+                Place {
+                    base: self.clone().base,
+                    elems: tcx.intern_place_elems(place_elems),
                 },
-                _ => None,
-            },
+                projection,
+            )),
             _ => None,
         }
     }
+
+    // for projection returns the base place;
+    //     Base.[a, b, c] => Base.[a, b]
+    //                 ^-- projection
+    // if no projection returns the place itself,
+    //     Base.[] => Base.[]
+    //          ^^-- no projection
+    pub fn base_place<'cx, 'gcx>(
```

Comment:
Probably a bad idea, because of inefficiency - you should be able to rewrite all the algorithms to work on `[PlaceElem]` instead of the recursive data structure.